src/System.Memory/ref/System.Memory.cs

```diff
@@ -112,6 +112,24 @@ public static partial class MemoryExtensions
 }
 namespace System.Buffers
 {
+    public sealed partial class ArrayBufferWriter : System.Buffers.IBufferWriter<byte>, System.IDisposable
```

Comment:
Should this type be generic?

Reply:
What would writing to the stream look like? We cast the `T[]` to a `byte[]`? Also, what contraints on T did you have in mind, if any? `T : unmanaged`?

Reply:
It should be generic, no constraint. Why wouldn't we make it generic? Not doing so closes of possible uses, for what gain?

> What would writing to the stream look like

Exactly the same as if it wasn't generic, except you instantiate an `ArrayBufferWriter<byte>` instead of an `ArrayBufferWriter`.

Are you referring to CopyToAndReset? I don't think we should have those at all. Just Write{Async} the OutputSpan/Memory to the stream and then call Reset. This not only simplifies the API and avoids issues with it being generic, it will likely also lead to better performance, because you won't force another state machine into existence just to call Reset.

Reply:
I initially didn't realize we were exposing a single output buffer. I guess there's no need to have a CopyTo at all since the caller has access to the underlying buffer..

Reply:
The reason for introducing `CopyTo` was that we wanted to avoid putting the caller in a state where they copy the data to the output stream but do not call `Clear()` afterwards. With such a helper, the caller wouldn't end up in that state (where they start duplicating the data writes).